lib/xnvme_be_macos_driverkit_dev.c

```diff
@@ -0,0 +1,218 @@
+	tmp_opts.nsid = 1;
+
+	while (true) {
+		struct xnvme_dev *dev = NULL;
+		char uri[XNVME_IDENT_URI_LEN] = {0};
+		io_service_t service = IOIteratorNext(iterator);
+
+		if (!service) {
+			break;
+		}
+
+		ret = IORegistryEntryGetName(service, uri);
+		if (strncmp("MacVFN-", uri, 7)) {
+			continue;
+		}
+		XNVME_DEBUG("MacVFN uri matched: %s", uri);
+
+		dev = xnvme_dev_open(uri, &tmp_opts);
+		if (!dev) {
+			XNVME_DEBUG("xnvme_dev_open(): %d", errno);
+			return -errno;
+		}
+		if (cb_func(dev, cb_args)) {
+			xnvme_dev_close(dev);
+		}
+	}
+
+	IOObjectRelease(iterator);
+
+	return -ENOSYS;
+}
+
+void
+xnvme_be_macos_driverkit_dev_close(struct xnvme_dev *dev)
+{
+	struct xnvme_be_macos_driverkit_state *state =
+		(struct xnvme_be_macos_driverkit_state *)dev->be.state;
+
+	if (!state->device_service) {
+		return;
+	}
+
+	IOServiceClose(state->device_service);
+	IOServiceClose(state->device_connection);
+}
+
+int
+_xnvme_be_driverkit_create_ioqpair(struct xnvme_be_macos_driverkit_state *state, int qd, int flags)
+{
+	NvmeQueue ioqpair_return;
+	kern_return_t ret;
+	size_t output_cnt;
+	unsigned int qid = __builtin_ffsll(state->qidmap);
+	// int err;
+
+	if (qid == 0) {
+		XNVME_DEBUG("no free queue identifiers\n");
+		return -EBUSY;
+	}
+
+	qid--;
+
+	XNVME_DEBUG("nvme_create_ioqpair(%d, %d, %d)", qid, qd + 1, flags);
+
+	// nvme queue capacity must be one larger than the requested capacity
+	// since only n-1 slots in an NVMe queue may be used
+	NvmeQueue ioqpair_create = {
+		.id = (uint64_t)qid,
+		.depth = (uint64_t)qd + 1,
+		.vector = -1,
+		.flags = (uint64_t)flags,
+	};
+	output_cnt = sizeof(NvmeQueue);
+
+	ret = IOConnectCallStructMethod(state->device_connection, NVME_CREATE_QUEUE_PAIR,
+					&ioqpair_create, sizeof(NvmeQueue), &ioqpair_return,
+					&output_cnt);
+	if (ret != kIOReturnSuccess) {
+		XNVME_DEBUG("NVME_CREATE_QUEUE_PAIR failed with error: 0x%08x, %s.\n", ret,
+			    mach_error_string(ret));
+	}
+	XNVME_DEBUG("NVME_CREATE_QUEUE_PAIR done\n");
+	if (ret) {
+		return -EAGAIN;
+	}
+
+	state->qid_sync = ioqpair_return.id;
+	state->qidmap &= ~(1ULL << qid);
+
+	return qid;
+}
+
+int
+_xnvme_be_driverkit_delete_ioqpair(struct xnvme_be_macos_driverkit_state *state, int qid)
+{
+	size_t output_cnt = 0;
+	NvmeQueue ioqpair_delete = {
+		.id = (uint64_t)qid,
+	};
+
+	kern_return_t ret =
+		IOConnectCallStructMethod(state->device_connection, NVME_DELETE_QUEUE_PAIR,
+					  &ioqpair_delete, sizeof(NvmeQueue), NULL, &output_cnt);
+	if (ret != kIOReturnSuccess) {
+		XNVME_DEBUG("NVME_DELETE_QUEUE_PAIR failed with error: 0x%08x, %s.\n", ret,
+			    mach_error_string(ret));
+	}
+	XNVME_DEBUG("NVME_DELETE_QUEUE_PAIR done\n");
+
+	state->qidmap |= 1ULL << qid;
+
+	return 0;
+}
+
+int
+xnvme_be_macos_driverkit_dev_open(struct xnvme_dev *dev)
+{
+	struct xnvme_be_macos_driverkit_state *state =
+		(struct xnvme_be_macos_driverkit_state *)dev->be.state;
+	int qpid;
+
+	kern_return_t err;
+	io_service_t service;
```

Comment:
See notes on declaration / initialization order and error-handling.

Reply:
Done.